contracts/staking/EpochSelector.sol

```diff
@@ -0,0 +1,1683 @@
+    /// @return Returns the new root after the rotation
+    function _rightRotate(address addressOfZ) internal returns (address) {
+        if (addressOfZ == address(0)) {
+            revert(ClusterLib.CANNOT_RR_ADDRESS_ZERO);
+        }
+        Node storage z = nodes[addressOfZ];
+
+        Node storage y = nodes[z.left];
+
+        // do not rotate if left is 0
+        if (y.node == address(0)) {
+            // console2.log("RR: not because y is 0 ");
+            return z.node;
+        }
+        Node storage T3 = nodes[y.right];
+
+        // cut z.left
+        z.sumOfLeftBalances = _getTotalBalancesIncludingWeight(T3.node);
+        z.left = T3.node;
+        // cut y.right
+        y.sumOfRightBalances = _getTotalBalancesIncludingWeight(z.node);
+        y.right = z.node;
+
+        z.height = Math.max(height(z.right), height(z.left)) + 1;
+        y.height = Math.max(height(y.right), height(y.left)) + 1;
+        return y.node;
+    }
+
+    /// @notice Lef rotate a given node
+    /// @param addressOfZ address of the node to left rotate
+    /// @return Returns the new root after the rotation
+    function _leftRotate(address addressOfZ) internal returns (address) {
+        if (addressOfZ == address(0)) {
+            revert(ClusterLib.CANNOT_LR_ADDRESS_ZERO);
+        }
+        Node storage z = nodes[addressOfZ];
+
+        Node storage y = nodes[z.right];
+
+        // do not rotate if right is 0
+        if (y.node == address(0)) {
+            // console2.log("LR: not because y is 0 ");
+            return z.node;
+        }
+        Node storage T2 = nodes[y.left];
+
+        // cut z.right
+        z.sumOfRightBalances = _getTotalBalancesIncludingWeight(T2.node);
+        z.right = T2.node;
+        // cut y.left
+        y.sumOfLeftBalances = _getTotalBalancesIncludingWeight(z.node);
+        y.left = z.node;
+
+        z.height = Math.max(height(z.left), height(z.right)) + 1;
+        y.height = Math.max(height(y.left), height(y.right)) + 1;
+
+        return y.node;
+    }
+
+    /// @notice Returns the node balance i.e difference in heights of left and right nodes
+    /// @param node Address of the node to get balance of
+    /// @return Balance of the node
+    function getBalance(address node) public view returns (int256) {
+        if (node == address(0)) return 0;
+
+        Node memory existingNode = nodes[node];
+
+        return int256(height(existingNode.left)) - int256(height(existingNode.right));
+    }
+
+    /// @notice Returns the data of the node
+    /// @param _node Address of the node
+    /// @return node Data of the node
+    function nodeData(address _node) public view returns (Node memory node) {
+        node = nodes[_node];
+    }
+
+    /// @notice Get total weight of the node
+    /// @param _node Address of the node to calculate total weight for
+    /// @return Total weight of the node
+    function _getTotalBalancesIncludingWeight(address _node) internal view returns (uint256) {
```

Comment:
Can we reuse `_getIndexesWithWeights` function instead of this seperate function ?